core/src/main/java/org/apache/calcite/util/Sources.java

```diff
@@ -54,6 +60,19 @@ public static Source file(File baseDirectory, String fileName) {
     }
   }
 
+  /**
+   * Create {@link Source} from a generic text source such as string, {@link java.nio.CharBuffer}
+   * or text file. Useful when data is already in memory or can't be directly read from
+   * a file or url.
+   *
+   * @param source generic "re-redable" source of characters
+   * @return {@code Source} delegate for {@code CharSource} (can't be null)
+   * @throws NullPointerException when {@code source} is null
+   */
+  public static Source fromCharSource(CharSource source) {
```

Comment:
Is this significantly better than `java.io.Reader` and/or `java.io.InputStream`?

Reply:
Yes. The idea is that `Reader` / `InputStream` are "single-pass" in the sense that you can't re-read them. CharSource is a supplier of `Reader` similarly to File / URL. 

You can't provide InputStream to Source because it can be read only once.

Reply:
Thanks, I see. That is fine then.